### DISCUSSION TITLE
pip, pipenv, poetry. What is your opinion?

### DISCUSSION THREAD

**Discussion by antonpirker** [getsentry/sentry-python#1620] [Features / New Integrations / Ideas]
To have faster tests in our CI run, we would need something like a requirements lock file.
See also this issue: https://github.com/getsentry/sentry-python/issues/1618

What is the communities opinion on those "package managers"? (are they called like that?)

Is it cool or not, will this be the standard in the future or just a waste of time? Let's talk!

**Comment by asottile-sentry** [comment]
a library shouldn't have a lock file -- you want to know as soon as possible when your dependencies break things.  lock files are for _applications_

**Reply by antonpirker** [reply]
I meant just for caching the virtual environments that are used for running our tests. 
I think we just would need to create a lock file while running CI so we can calculate a cache from that lock file, so if then a new versions of something in the lock file is released between two test runs, the cached virtualenv for running the tests in is invalidated and created anew

**Reply by asottile-sentry** [reply]
yes that's what I mean -- if you have a lock file for tests then you aren't going to be testing the latest versions -- you'll be testing the locked versions

**Comment by alexted** [comment]
A few years ago I had the great pleasure of giving up using vanilla pip in favor of pipenv - it was super, I couldn't be happier! Working with dependencies in the project became very pleasant and systematic. After a couple of years, pipenv stopped making me happy because of bugs (no ability to update one particular library, instead it updates all of them) and very low performance (dependency resolution and locking libraries versions takes VERY long), so a one year ago I decided it was time to switch to poetry - and, you know, since then I have never regretted my decision once. So, in my opinion, intelligent dependencies managers are the future, and they already have huge benefits.
The proof of this I believe is as follows: 
https://peps.python.org/pep-0517/
https://peps.python.org/pep-0518/
https://peps.python.org/pep-0680/
